src/check_ast.rs

```diff
@@ -2009,6 +2021,7 @@ where
                 }
             }
         }
+        self.pop_scope();
```

Comment:
I'd prefer to find a way to implement this that doesn't require introducing a new scope, since the scope semantics are kinda tricky. I'd even prefer adding an `in_excepthandler: bool` to the `Checker` if that's easiest :)

Reply:
Got it. How about implementing a visitor that looks like this?

```rust
impl<'a> Visitor<'a> for RaiseVisitor {
    fn visit_stmt(&mut self, stmt: &'a Stmt) {
        match &stmt.node {
            StmtKind::Raise { exc, cause } => {
                if cause.is_none() {
                    if let Some(exc) = exc {
                        match &exc.node {
                            ExprKind::Name { id, .. } if is_lower(id) => {}
                            _ => {
                                self.checks.push(Check::new(
                                    CheckKind::RaiseWithoutFromInsideExcept,
                                    Range::from_located(stmt),
                                ));
                            }
                        }
                    }
                }
            }
            StmtKind::FunctionDef { .. }
            | StmtKind::AsyncFunctionDef { .. }
            | StmtKind::ClassDef { .. } => {}
            StmtKind::Try { .. } => {}
            StmtKind::If { body, .. }
            | StmtKind::With { body, .. }
            | StmtKind::AsyncWith { body, .. }
            | StmtKind::For { body, .. }
            | StmtKind::AsyncFor { body, .. }
            | StmtKind::While { body, .. } => {
                for stmt in body {
                    self.visit_stmt(stmt);
                }
            }
            _ => {}
        }
    }
}
```

Reply:
> I'd even prefer adding an in_excepthandler: bool to the Checker if that's easiest :

Can you elaborate on this approach?

Reply:
Yeah that looks reasonable to me.

What I was suggesting was something like:

```rust
fn visit_excepthandler(&mut self, excepthandler: &'b Excepthandler) {
    prev_in_excepthandler = self.in_excepthandler;
    self.in_excepthandler = true;
    ...
    self.in_excepthandler = prev_in_excepthandler;
}
 ```

Then here, you could check `if self.in_excepthandler`. It's kind of a mess because we have to do that separate bookkeeping, but it does mean to can avoid multiple passes over the AST as would be required with the visitor.

Reply:
Does this approach work when a raise statement is in a function?

```python
except:
   def foo():
       raise Exception
```

Reply:
Probably not, no!

Reply:
(So I think the visitor you outlined above is probably a better choice.)

Reply:
Got it :) I'll push the code tonight.

Reply:
Thanks!